apps/sms/services/js/activity/activity_shim.js

```diff
@@ -44,30 +47,45 @@
     /**
      * Reference to currently active activity request.
      * @type {ActivityRequestHandler}
+     * @private
      */
     [priv.request]: null,
 
     /**
-     * Initialized activity service bridge.
+     * Reference to mozSetMessageHandler function.
+     * @type {Function}
+     * @private
+     */
+    [priv.setMessageHandler]: null,
+
+    /**
+     * Initialized activity shim bridge service.
+     * @param {string} appInstanceId Unique identifier of app instance where
+     * shim resides in.
      */
-    init() {
-      this.initService();
-
-      // We use setTimeout here to allow activity client to do connection
-      // handshake with the service first, so that it will be able to receive
-      // event broadcasted by service.
-      setTimeout(() => {
-        navigator.mozSetMessageHandler(
-          SYSTEM_MESSAGE_NAME, this[priv.onActivityRequest].bind(this)
-        );
-      });
+    init(appInstanceId, setMessageHandler) {
+      if (!appInstanceId) {
+        throw new Error('AppInstanceId is required!');
+      }
+
+      this[priv.setMessageHandler] = setMessageHandler;
+
+      this.initService(
+        new BroadcastChannel(`${SERVICE_NAME}-channel-${appInstanceId}`)
```

Comment:
Ok, it seems like we should migrate all the direct window - iframe communication to this instead of BroadcastChannel. Maybe we can create a bug and leave a comment here?